pilot/info/storagekeys.py

```diff
@@ -0,0 +1,103 @@
+            else:
+                logger.info("Got key pair returns wrong value: %s" % res)
+        except Exception as ex:
+            logger.error("Failed to get key pair(%s,%s): %s, %s" % (access_key, secret_key, ex, traceback.format_exc()))
+        return {}
+
+    def get_special_setup(self, ddmendpoint, protocol_id):
+        """
+        Construct special setup for ddms such as objectstore
+        :param ddmendpoint: storage data
+        :param protocol_id: protocol id.
+        :return: setup as a string
+        """
+
+        logger.info("Get special setup for ddmendpoint(%s) protocol id(%s)" % (ddmendpoint.name, protocol_id))
+        if protocol_id is None or str(protocol_id) not in ddmendpoint.rprotocols.keys():
+            return None
+
+        if ddmendpoint.name in self.storage_keys and protocol_id in self.storage_keys[ddmendpoint.name]:
+            return self.storage_keys[ddmendpoint.name][protocol_id]['special_setup']
+
+        if ddmendpoint.type in ['OS_ES', 'OS_LOGS']:
+            if ddmendpoint.name not in self.storage_keys:
+                self.storage_keys[ddmendpoint.name] = {}
+            if protocol_id not in self.storage_keys[ddmendpoint.name]:
+                self.storage_keys[ddmendpoint.name][protocol_id] = {}
+            self.storage_keys[ddmendpoint.name][protocol_id]['special_setup'] = None
+
+            settings = ddmendpoint.rprotocols.get(str(protocol_id), {}).get('settings', {})
+            access_key = settings.get('access_key', None)
+            secret_key = settings.get('secret_key', None)
+            is_secure = settings.get('is_secure', None)
+
+            # make sure all things are correctly defined in AGIS.
+            # If one of them is not defined correctly, will not setup this part. Then rucio client can try to use signed url.
+            # This part is preferred because signed url is not efficient.
+            if access_key and secret_key and is_secure:
+                key_pair = self.get_security_key(secret_key, access_key)
+                if "privateKey" not in key_pair or key_pair["privateKey"] is None:
+                    logger.error("Failed to get the key pair for S3 objectstore from panda")
+                else:
+                    setup = "export S3_ACCESS_KEY=%s; export S3_SECRET_KEY=%s; export S3_IS_SECURE=%s;" % (key_pair["publicKey"],
+                                                                                                           key_pair["privateKey"],
+                                                                                                           is_secure)
+                    self.storage_keys[ddmendpoint.name][protocol_id]['special_setup'] = setup
+                    logger.info("Return key pair with public key: %s" % key_pair["publicKey"])
+                    return self.storage_keys[ddmendpoint.name][protocol_id]['special_setup']
+        return None
+
+
+storage_keys = StorageKeys()
```

Comment:
Wen,
 I'm always wondering when someone uses "singletone" objects in python as global/module variable. If an object is instantiated in global namespace of a module, it automatically becomes singletone-based, since it's global, right?

 so  Singletone-base and global/module variable do same business, why it's needed to use them together?

did I miss something?

Reply:
Yes. It's duplicated.
At first, I didn't plan to use global variable. So singleton function wad added. But then I found it's too complicated to initialize it at different place. So I removed different initialization at different place and put it as a global variable.

Reply:
Wen, actually the same comment about using singleton as a cache for StorageKeys,
for me this is again redundant, you can always resolve it directly from `StorageData` which is available from cache-able `infosys`.

Sending requests to Panda can be integrated as data provider for storage data.

Reply:
It was in storagedata in last pull request too.